### ISSUE TITLE
`__array__` protocol not properly supported

### ISSUE THREAD

**Issue by bayerj** [google/jax#3092]
**tl;dr:** Overloading `__array__` does not result in the same behaviour for `jax.numpy` and `numpy`. 

This behaviour is nice to create classes that have an additional methods, but which should behave as an array. 

### Problem description

Assume I overload the numpy specific magic method `__array__` in the following container:

```python
class ArrayContainer:
  def __init__(self, value):
    self.value = value
  def __array__(self):
    return self.value
```

Using this with a numpy function, such as concatenate, will work nicely by unpacking the contained value via `__array__()`:

```python
container = ArrayContainer(onp.zeros((2)))
print(np.concatenate([container, container], -1)).   # prints [0. 0. 0. 0.]
```

However, this raises a `ValueError` in jax:

```python
jax_container = ArrayContainer(jnp.zeros((2)))
print(jnp.concatenate([jax_container, jax_container], -1)).  # ValueError: object __array__ method not producing an array
```

The complete example and the full traceback can be found [in this gist.](https://gist.github.com/bayerj/509858e15aa998c43ddb40b33c56dbbc)

**Comment by jakevdp** [comment]
I believe this is happening for the same reason that jax routines don't work with lists:
```python
>>> import jax.numpy as jnp
>>> jnp.concatenate([[1, 2, 3], [4, 5, 6]])
TypeError: data type not understood
```
In general, jax does not promote arbitrary inputs to jax arrays. If you want to use arbitrary inputs, the best approach is to manually wrap the inputs in ``jnp.asarray``:
```python
print(jnp.concatenate([jnp.asarray(jax_container), jnp.asarray(jax_container)], -1))
```

**Comment by bayerj** [comment]
But the error produced is different.

It also works with `jnp.array` instead of `asarray`. I just think it would be great if jax did this implicitly, just as numpy.
